### PR TITLE
Migration-Requirejs: view source files & diff apps

### PR DESCRIPTION
**Pull request by Rohit25negi**
https://trello.com/c/udnaooKF/36-1-corehq-apps-appmanager-quick-wins

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/js/download_index_main.js
Comment (Rohit25negi) [reply corehq/apps/app_manager/static/app_manager/js/download_index_main.js]: do we restrict use of for-of loops?

Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/download_index_main.js]: I believe this is the first time someone has used one in HQ. Given that stickler doesn't like it and that it isn't used elsewhere in the codebase, I'd switch this to `_.each`.

File: corehq/apps/app_manager/static/app_manager/js/source_files.js
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/source_files.js]: this should be safe to switch to `!==`

File: corehq/apps/app_manager/templates/app_manager/download_index.html
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/download_index.html]: I'm so glad to see this going away.

File: corehq/apps/app_manager/static/app_manager/js/download_index_main.js
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/download_index_main.js]: `ace` should be added to params here since it's accessed below

File: corehq/apps/app_manager/static/app_manager/js/app_diff_main.js
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_diff_main.js]: Since this only has one dependency, you can just use `app_manager/js/source_files` as the `requirejs_main` modules for `app_diff.html` and delete this file.

**Review by orangejenny — CHANGES_REQUESTED** [review]
Minor changes requested.

File: corehq/apps/app_manager/static/app_manager/js/download_index_main.js
Comment (Rohit25negi) [reply corehq/apps/app_manager/static/app_manager/js/download_index_main.js]: ohh shit. Stupid mistake :triumph: . I should have been more careful.

**Comment by orangejenny** [comment]
Hey, about PR titles: other people look at them for a number of reasons: blaming a file or doing something in `git log` (possibly months or years in the future); skimming titles in the deploy email to guess which PR caused a bug that showed up immediately after a deploy; reading them to put together a summary of recent changes for global services or external users.

The people reading the titles might be other devs or might be on support or product. So it's worth making the title reader-friendly by the time the PR gets merged: removing your initials, removing "WIP", and using non-dev-centric descriptions where possible. I'd call this PR something like "requirejs migration: app manager: view source files & diff apps" - it tells future devs *which* migration, and it's understandable by non-devs who don't know what the "download index" page is.

**Comment by Rohit25negi** [comment]
@orangejenny PR is good to go once approved. 
Resolved the ace-mode loading problem we talked about.

File: corehq/apps/app_manager/static/app_manager/js/download_index_main.js
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/download_index_main.js]: This doesn't need to go inside the document ready, right? So it could be
```
hqRequire([...], function() {
   $(function () {
      ...
   });
})
```
and potentially start loading the ace modes before the document is done loading.

**Comment by orangejenny** [comment]
> Resolved the ace-mode loading problem we talked about.

This looks like it was quite a bear to track down, nice. This issue is going to happen on any requirejs page that uses ace, right? Right now that's `hqadmin/js/admin_restore`, `hqadmin/js/raw_couch_main`, and `userreports/js/base`. Could you take a look at using the ace logic in those places?

Since `userreports/js/base` is already all ace-related logic, it probably makes the most sense to rename that to something like `hqwebapp/js/ace`, extend it to handle modes, and use it on the other ace pages.

**Comment by Rohit25negi** [comment]
> > Resolved the ace-mode loading problem we talked about.
> 
> This looks like it was quite a bear to track down, nice. This issue is going to happen on any requirejs page that uses ace, right? Right now that's `hqadmin/js/admin_restore`, `hqadmin/js/raw_couch_main`, and `userreports/js/base`. Could you take a look at using the ace logic in those places?
> 
> Since `userreports/js/base` is already all ace-related logic, it probably makes the most sense to rename that to something like `hqwebapp/js/ace`, extend it to handle modes, and use it on the other ace pages.

Hey please do not merge this PR, while making base  module for ace. Understood that hqRequire is making the code complex and unnecessarily lengthy(requires to specify modules separately). I am testing it with setting the ace basePath. After a sanity test for this and 'base_ace' module that I will make, this should be merged.

**Comment by orangejenny** [comment]
@Rohit25negi Cool, got it. Added the "Open for review: do not merge" tag back to make it less likely this gets carelessly merged (since the required-labels check is now failing).

**Review by orangejenny — APPROVED** [review]

**Pull request merged**
